### PR TITLE
Fix current alerts focus styles - part 2

### PR DESCRIPTION
**Pull request by tombye**
Fix for the large links on the current/past/rejected alerts pages, which currently overlap content below them:

current_alerts_before

Part of https://www.pivotaltracker.com/story/show/180011155.

This uses JavaScript to apply styles that make the content poke through the overlap and give them the right colour contrast.

current_alerts_after_with_js

## What caused the bug

We have CSS for this style of focus (see the templates page for an example) but the version on the current/past/rejected alerts pages has a slightly different HTML which means this CSS cannot target it.

These kind of links are normally made up of this kind of HTML:

```html
<li>
  <a>Title of item</a>
  <span>Description of it / more info about it</span>
</li>
```

So we can use a CSS selector like this to target the `<span>` when the link is focused:

`a:focus + span` (match a 'span' which is the next sibling of a focused 'a').

But these ones are more like:

```html
<div>
  <h2>
    <a>Title of item</a>
  </h2>
  <span>Description of it / more info about it</span>
  <span>Status of it</span>
</div>
```

Because the `<a>` is (correctly) wrapped in a `<h2>`, we can't target the `<span>`s as its siblings so our selector doesn't work. As far as I know, there's also no way you can currently write a selector for this HTML*. So we have to add a class to the `<h2>` using JS.

*The [:has() pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:has) will let us do this but is only supported in Safari tech preview at the moment.

## Problems with using JS on content that is updated via AJAX

Unfortunately it's not as simple as that because the `<h2>` is part of a block of HTML controlled by [the AJAXy code](https://github.com/alphagov/notifications-admin/blob/master/app/assets/javascripts/updateContent.js) that sync's it with HTML sent from the server-side app. This means any classes added to it will be wiped by the sync'ing because they don't exist in the HTML from the server-side app.

## The proposed solution

So this proposes the following solution:
- changes to the AJAXy code so it allows **some** classes to be ignored from comparisons with the server-side HTML and to persist through any updates
- some JS, that runs when the links are focused and unfocused, to add/remove a class on the `<h2>`

## Notes for reviewers

These changes are intended to be read commit-by-commit. The commits have more detail on the changes and the sequence is meant to make things a bit more digestible.

### CODE REVIEW

**Comment by tombye** [comment]
Added a guard to some code here that operates on DOM nodes after talking with @benthorner made it clear it assumes those nodes exist incorrectly.

File: app/assets/javascripts/updateContent.js [hunk 1]
```diff
@@ -84,6 +84,14 @@
...
     this.start = component => {
       var $component = $(component);
+      var $contents = $component.children().eq(0);
+      var key = $component.data('key');
+      var resource = $component.data('resource');
+      var form = $component.data('form');
+
+      // replace component with contents
+      // the renderer does this anyway when diffing against the first response
+      $component.replaceWith($contents);
```
Comment (benthorner) [inline app/assets/javascripts/updateContent.js:98]: (As discussed)

Rather than doing this, could we just pass morphdom the contents? Some advantages of doing it that way:

- It's simpler - a one line change / fix.
- It preserves the outer div to match the source HTML.

I understand this approach is more conservative - it just accelerates the current behaviour - but I'm confident it's safe to go one step further and keep the outer div everywhere. In particular, if something is relying on the outer div being absent, then I doubt it works properly without JS (where the div still exists). In addition, while this change is conservative, it's still possible that something is relying on the initial change being delayed.

What do you think?

Related to this is your note about removing the (currently redundant) aria attribute. While this is indeed clearer, could we give screen reader users a better experience if we were able to keep it?

Optional: irrespective of what we agree, it could be helpful to future devs to have a test that documents the behaviour.

File: app/assets/javascripts/updateContent.js [hunk 2]
```diff
@@ -26,7 +26,11 @@
     },
     replace: function () {
       this.classNames.forEach((className, index) => {
-        this.$els[index].addClass(className);
+        var $el = this.$els[index];
+
+        if (global.document.body.contains($el.get(0))) {
```
Comment (benthorner) [inline app/assets/javascripts/updateContent.js:31]: Are you sure this is necessary? I made a quick tweak to one of the new JS tests - the one that removes the second list item - and didn't get any errors without this commit. Might be worth having a test to cover this if it is necessary.

**Review by benthorner — COMMENTED** [review]
Great explanations as usual 💯. Thanks also for spending a bunch of time talking through some of the changes and my questions about them 🏅 . A few notes from that:

- The majority of the (JS) changes in this PR are in order to make the hint text black - instead of grey - when an alert is focussed - the grey version is the JS fallback, which is pretty good considering it's all we can do until the `has` pseudoclass comes along.

- The easiest way of fixing the bug would actually be to remove the outer `h2`, which would be consistent with how we markup templates, for example. However, I agree that what we have here is semantically more correct and helpful than styled `<a>`s.

Short version: this is the least worst solution and I admire your tenacity in pursuing it.

I've left a comment about one thing to fix before this goes in. At the same time, I'm keen to see if we can cut down some of complexity of the diffing code and its tests.

## Screenshot of the no-JS fallback

Screenshot 2022-02-11 at 14 33 21

File: app/assets/javascripts/updateContent.js (hunk 2 above)
Comment (tombye) [reply app/assets/javascripts/updateContent.js:31]: It wouldn't error because when DOM nodes are removed they aren't automatically deleted, just detached from the DOM. I just figured that without it the code supports a path whereby nodes no longer part of the DOM are still operated on, which is a bit of red-herring if you're debugging this block.

Given this change is more about how it reads, do you think there's value in a test for that case?

File: app/assets/javascripts/updateContent.js (hunk 1 above)
Comment (tombye) [reply app/assets/javascripts/updateContent.js:98]: I'm still a little nervous about changing the HTML left in the page across all our AJAX code, at least without checking all the uses, so think it's out of scope for this PR. I like your suggestion to add a test for this version so did (in https://github.com/alphagov/notifications-admin/pull/4155/commits/51594e7d461922bd6a7d2d7828d641c1a5613e2a) and will have a think about whether to move the change into the `render` method separately.

File: app/assets/javascripts/updateContent.js [hunk 3]
```diff
@@ -11,10 +11,37 @@
       1000
   ));
 
-  var getRenderer = $component => response => morphdom(
-    $component.get(0),
-    $(response[$component.data('key')]).get(0)
-  );
+  // Methods to ensure the DOM fragment is clean of classes added by JS before diffing
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/javascripts/updateContent.js:14]: I agree that code related to the focus class should be marked as something that should (will?) go away at some point but this object was written to be intentionally generic from whatever JS is operating on the updateContent HTML. That being the case, do you still think the comment should live here or could it be part of a comment above the code in `app/assets/javascripts/main.js` that controls the focus class?

File: app/assets/javascripts/updateContent.js [hunk 4]
```diff
@@ -11,10 +11,37 @@
...
+  // and that they are replaced afterwards.
+  var classesToPersist = {
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/javascripts/updateContent.js]: Great suggestion. I've changed it in https://github.com/alphagov/notifications-admin/pull/4155/commits/4f59396a0017e70168ef3aaf6a033d24c2df2252.

File: app/assets/javascripts/updateContent.js [hunk 5]
```diff
@@ -55,15 +82,26 @@
...
-        $(component).data('form')
-      ),
-      defaultInterval
-    );
+    this.start = component => {
+      var $component = $(component);
+
+      // store any classes that should persist through updates
+      if ($contents.data('classesToPersist') !== undefined) {
+        $contents.data('classesToPersist')
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/javascripts/updateContent.js:105]: Good point, and the other methods use it in ways that having it public could break. Made it private (at least by naming and implied access) in https://github.com/alphagov/notifications-admin/pull/4155/commits/9646e17b033b5107697cb578cff23bd88a527aab.

File: tests/javascripts/updateContent.test.js [hunk 6]
```diff
@@ -50,202 +50,438 @@ describe('Update content', () => {
...
-          window.GOVUK.Modules.UpdateContent.calculateBackoff(responseTime)
-        ).toBe(
-          waitTime
-        );
     });
 
   });
 
-  describe("If a form is used as a source for data, referenced in the data-form attribute", () => {
+  describe("When adding or removing DOM nodes", () => {
```
[the remark this answers is not in the document]
Comment (tombye) [reply tests/javascripts/updateContent.test.js:253]: I'm not totally sure I understand this comment as well as I should. Do you mean the HTML fragments used should be simplified to be made up rather than taken from the app, as they are?

When you mention this being separate from the HTML near the top, do you think we should have one HTML string and swap out the partial in it to set up testcases for HTML that:
- has its content updated
- has its elements added/removed
?

I think it is a bit crunky to read as I did just bolt some new code onto the existing tests rather than re-evaluating how the whole thing reads. Be interested in your thoughts on this.

File: app/assets/stylesheets/views/dashboard.scss [hunk 7]
```diff
@@ -110,13 +115,6 @@
 
 /* The focus state for sibling links overlaps the hint so the hint's text colour needs to adapt */
 .govuk-link:focus {
-  &.file-list-filename,
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/stylesheets/views/dashboard.scss:113]: After a bit of testing I actually think `.file-list-filename` and `.file-list-filename-large` didn't need to have `position: relative` so it makes sense for them to not be included in the classes moved to block further up. Those classes are always applied to the large links so don't ever get overlapped by the focus style*.

*For clarity, I mean that we were applying `position: relative` to elements to make them 'positioned' and so have a higher staking order (see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/Stacking_without_z-index) than the overlapping anchor element.

I also can't replicate what you're seeing in the screenshot, which looks like the style before these changes, but I'm guessing here. Could you confirm it's still happening please?

File: tests/javascripts/updateContent.test.js [hunk 8]
```diff
@@ -50,202 +50,438 @@ describe('Update content', () => {
...
+        }
+      ]);
+
+      initialHTMLString = `<div data-module="update-content" data-resource="${resourceURL}" data-key="${updateKey}" aria-live="polite">
+                          ${HTMLString}
+                          </div>`;
+
+      document.body.innerHTML = initialHTMLString;
+
+      // make the response have an extra item
```
[the remark this answers is not in the document]
Comment (tombye) [reply tests/javascripts/updateContent.test.js]: Agreed. Should be fixed in https://github.com/alphagov/notifications-admin/pull/4155/commits/e310ff34696d0f280bbfd45355eed50bc8650cff

File: tests/javascripts/updateContent.test.js [hunk 9]
```diff
@@ -50,202 +50,438 @@ describe('Update content', () => {
...
-      ]));
 
-    })
+      // check the node has been removed
+      expect(document.querySelectorAll('.file-list').length).toEqual(1);
+      expect(document.querySelectorAll('.file-list h2 a')[0].textContent.trim()).toEqual("Gas leak");
+
+    });
+
+    test("If other scripts have added classes to the DOM, they should persist through updates", () => {
```
[the remark this answers is not in the document]
Comment (tombye) [reply tests/javascripts/updateContent.test.js:444]: Good point! I've changed the test to actually update something in https://github.com/alphagov/notifications-admin/pull/4155/commits/e310ff34696d0f280bbfd45355eed50bc8650cff

File: app/assets/javascripts/updateContent.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:14]: Hmm...potentially both. My concern is that it will get reused by accident and then we won't be able to remove it when we're ready i.e. we should discourage other uses of it if a reasonable alternative exists.

I don't feel strongly about this, though. On a practical note, the most likely person to change this stuff is you, and you'll remember. That said, how will we keep track of when `has()` is available?

That seems like a good reason for a comment in `main.js`, at least.

File: app/assets/javascripts/updateContent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:105]: Sorry, I don't think I was clear: my point was that multiple uses of `data-classes-to-persist` with different classes will get combined (or duplicated, if they're the same) in the global `classNames` property. One way to get around this would be to create a fresh copy of classPersister for each usage of updateContent.

Looking again, there's a general problem if this code is used with multiple class names: it'll re-add them to all elements, even if they weren't there before. One way to get around this would be to use a map instead of an array - to record which classes went with which elements.

I don't think either of these are a problem right now but may surprise someone who comes to reuse this. 

I'll leave it up to you if you want to fix either of them.

File: tests/javascripts/updateContent.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/javascripts/updateContent.test.js:444]: I'm still slightly wary of this: adding a new element - the approach in e310ff34696d0f280bbfd45355eed50bc8650cff - is still a bit different to changing the existing one, which is where we're trying to preserve the class; it's unclear if the DOM of the first item has been touched at all by the diffing library, so the test could pass when it shouldn't.

That said, I tried removing the persistence and this test fails, so ✅.

File: tests/javascripts/updateContent.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/javascripts/updateContent.test.js:253]: Yes please to the first paragraph. The bit that frustrated me was when I realised we don't actually use `areas` for anything. We could just hard-code those with a dummy value, for example.

Sort of yes to the second paragraph. I like the approach you've taken with the new code, which you could reuse for all the above tests - with a few tweaks to the selectors and assertions to match.

In particular, instead of dashboard summary stats and `HTMLString.replace(/<div class="big-number-number...` the test would use areas HTML from the generator functions you added.

Does that help?

File: app/assets/stylesheets/views/dashboard.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/stylesheets/views/dashboard.scss:113]: Ah, sorry, I was getting confused: using `position: relative` _creates_ that issue.

It does fix the non-clickable issue in the places this PR is concerned with, though. I'm not fussed if we fix that here, though, as it was an issue before this PR and remains an issue elsewhere.

File: app/assets/javascripts/updateContent.js (hunk 1 above)
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:101]: Fair enough. I'm not worried myself, but that's not based on anything objective.

I'm still curious, though: do you think the aria attribute would add value if it did actually work?

Comment (tombye) [reply app/assets/javascripts/updateContent.js:101]: I think, at this stage at least, the `aria-live` would add confusion because any users that would hear those announcements would get an update any time things changed. That seems bad because:
- they never made an active choice (via settings) or performed an interaction that asked for that information
- high-volume services would produce an almost unmanagable amount of noise with values changing all the time

In other words, I think we put it on originally with the intention that changes wouldn't be hidden from non-visual usage (which is a good intention) but maybe didn't see what they actually resulted in, partly because it didn't work so seemed pretty quiet.

In other words, this is something we should look into at some point but having it off means announcements aren't happening in a way that feels out of a user's control and you can still get the latest value by going to that part of the page when you need it, which I think is kind of how most visual users find that info' too.

Comment (tombye) [reply app/assets/javascripts/updateContent.js:101]: > they never made an active choice (via settings) or performed an interaction that asked for that information

Thinking about it, I'm not sure this is actually not strictly true all the time. When you send an email/sms, for example, the page you are redirected to has an updating region which switches from "Sending" to "Delivered" when sent. I'd argue that users who stay on that page might well want to know exactly when the email/sms was delivered. That flow was tested by the Digital Accessibility Centre in our audit 2 years ago and didn't produce any issues so I am guessing here.

I don't think this should block this PR but, seeing as we're having the conversation here, I'd be interested in what @quis thinks as this seems strongly linked to the intention of the design.

File: app/assets/javascripts/updateContent.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/javascripts/updateContent.js:14]: I've added some comments in https://github.com/alphagov/notifications-admin/pull/4155/commits/776f210e7b64e088d7ffaa8527c9d87c67aab9b8 to at least mark it for possible deletion.

I know I'm the main person who touches this code but others do bump into it so might need to know at least that it might not be needed. That and the possibility of me being hit by a bus, etc... As for tracking the availability of `has()`, that's trickier because we'd need to have dropped support for all non-evergreen browsers too, or at least decided to serve them something without JS to use `has()` as the only solution so it probably won't happen for a bit but could be triggered by work in any of those areas.

File: app/assets/javascripts/updateContent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/javascripts/updateContent.js:117]: I think there's value in at least checking for duplicates so I put some changes into https://github.com/alphagov/notifications-admin/pull/4155/commits/467be3a3394f7e46e6740c2fbb7061c7abeddc17. It's not exactly precise but, I think does the job for what we need this code to do at this point.

>Looking again, there's a general problem if this code is used with multiple class names: it'll re-add them to all elements, even if they weren't there before.

I'm not I understand this one. If it's about the same class names being added to an element multiple times, the `.addClass` method guards against this. Is that what you meant?

File: app/assets/javascripts/updateContent.js (hunk 2 above)
Comment (tombye) [reply app/assets/javascripts/updateContent.js:42]: Browsers all massively optimised their DOM access a while ago so I think it's hard to say for certain if it is expensive and hard to test in a simple way across the different browser engines. I do think this guard has value in how it makes clear that we're only wanting to add the class to elements in use and saves confusion if debugging.

All that being the case, I've added a comment (in https://github.com/alphagov/notifications-admin/pull/4155/commits/467be3a3394f7e46e6740c2fbb7061c7abeddc17) which kind of says what yours does but expands a bit on what an element being present means.

File: app/assets/javascripts/updateContent.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:14]: > we'd need to have dropped support for all non-evergreen browsers too

No one suspects IE (at least I don't). Sigh.

Thanks for pointing that out - this is clearly more permanent than I realised. I think the comments are still helpful in that they connect the two separate bits of code together more clearly 💯.

File: app/assets/stylesheets/views/dashboard.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/stylesheets/views/dashboard.scss:113]: >It does fix the non-clickable issue in the places this PR is concerned with, though

Does this mean that you're seeing clicks on the text below the link activate the link?

I've done some more testing and I don't think it's `position: relative` that's stopping clicks on the text underneath the link from passing through to it.

If you put `pointer-events: none` on this text on the templates page, it makes clicks on it pass through. This doesn't work on the current/past/rejected alerts pages because the text is wrapped in the grid `<div>`s (I found this by testing without the`<div>`s).

Also: I chatted with @quis about whether this needed fixing and we think it does so I added [a comment to the story this PR relates to](https://www.pivotaltracker.com/story/show/180011155/comments/229612021) which I'll pick up after this PR.

File: app/assets/javascripts/updateContent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:117]: Let me try and explain with an example:

```
# current usage of this, no problem
<div class="ajax-block-container js-mark-focus-on-parent" data-classes-to-persist="js-child-has-focus">
...
</div>

# another usage in the same HTML
<div class="ajax-block-container js-mark-focus-on-parent" data-classes-to-persist="some-other-class js-child-has-focus">
...
</div>
```

At the moment it's not a problem, but seeing as this code is being written generically I thought it worth point out how it could break in a reuse scenario like the above: elements in both divs will get the `js-` class applied twice, and elements in the first div will also get `some-other-class`.

Your commit fixes the first problem 👍. 

Some possible solutions to the second one (which would also fix the first problem):

- Make the persister code non-generic: it only ever looks for the `has-focus` class. That removes a bunch of complexity and you could always make it generic in future if needed.

- Switch the persister to use a map or isolate the persister object to each use of the component, so that the two divs above get their own separate set of classes to persist.

Equally, it's not a problem right now, so arguably you don't even need that extra commit. It might be helpful to a future dev to indicate the problem if the new code is reused elsewhere on a page.

Does that make a bit more sense?

File: app/assets/javascripts/updateContent.js (hunk 1 above)
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:109]: Thanks @tombye ! Sorry to elongate the discussion on this PR (although it's not a blocking one).

It sounds like a useful question to ask - especially with the example you gave.

File: app/assets/stylesheets/views/dashboard.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/stylesheets/views/dashboard.scss:113]: > Does this mean that you're seeing clicks on the text below the link activate the link?

Yup. But while it may superficially fix the issue, I have no opinion on what the actual solution should be, though. Sounds like this is now in the realm of follow-up work 👍.

**Comment by tombye** [comment]
Pushed a small fix to the focus style in https://github.com/alphagov/notifications-admin/pull/4155/commits/1d69f7ab6104e3978d84ff63657f6e130d85ca59 to plug a gap @quis noticed:

## Before

gap_above_alert_focus_style_before

## After

gap_above_alert_focus_style_after

File: app/assets/javascripts/updateContent.js (hunk 1 above)
Comment (tombye) [reply app/assets/javascripts/updateContent.js:109]: I've written up a story for this: https://www.pivotaltracker.com/story/show/181280213. I've purposely written it as a spike because the idea that updates should be announced by wrapping the partial in a live region is an assumption.

**Comment by benthorner** [comment]
Side note: I think this PR fixes a bug I just spotted.

The effect is even more pronounced with search (currently displays "Only showing the first 50 messages" twice).

## Before (master, after waiting a second or two)

| screenshot | elements |
| - | - |
| Screenshot 2022-02-16 at 10 56 49 | Screenshot 2022-02-16 at 10 56 57 |

## after (this branch)

| screenshot | elements |
| - | - |
| Screenshot 2022-02-16 at 10 57 03 | Screenshot 2022-02-16 at 10 57 43 |

File: app/assets/javascripts/updateContent.js (hunk 1 above)
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:109]: Nice - great summary on the card / comment 💯.

File: tests/javascripts/updateContent.test.js [hunk 10]
```diff
@@ -50,202 +50,468 @@ describe('Update content', () => {
...
-                        </div>`;
+    test("It should replace the original HTML with that of the partial, to match that returned from AJAX responses", () => {
 
-    document.body.innerHTML = initialHTMLString;
+      // start the module
+      window.GOVUK.modules.start();
 
-    // default the response to match the content inside div[data-module]
-    responseObj[updateKey] = HTMLString;
+      expect(document.querySelector('.ajax-block-container').parentNode.hasAttribute('data-resource')).toBe(false);
```
[the remark this answers is not in the document]
Comment (tombye) [reply tests/javascripts/updateContent.test.js]: Now I'm using individual functions to generate the HTML and calling them multiple times I do feel a bit more confident the test will start with the right HTML so might leave this as is. If you feel the approach didn't change your comment, do say.

File: app/assets/javascripts/updateContent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/assets/javascripts/updateContent.js:125]: That looks good to me 👍.

You could actually revert part of https://github.com/alphagov/notifications-admin/commit/467be3a3394f7e46e6740c2fbb7061c7abeddc17 now the component is only used once.

File: tests/javascripts/updateContent.test.js [hunk 11]
```diff
@@ -47,205 +47,463 @@ afterAll(() => {
...
-        ['serviceNumber', serviceNumber]
-      ]));
 
-    })
+      // re-select in case nodes in partialsInPage have changed
+      partialsInPage = document.querySelectorAll('.ajax-block-container');
+
+      // check the classes are still there
+      expect(partialsInPage[0].querySelectorAll('.file-list h2')[0].classList.contains('js-child-has-focus')).toBe(true);
+      expect(partialsInPage[1].querySelectorAll('.file-list h2')[0].classList.contains('js-2nd-child-has-focus')).toBe(true);
```
Comment (benthorner) [inline tests/javascripts/updateContent.test.js:492]:
```suggestion
      expect(partialsInPage[1].querySelectorAll('.file-list h2')[0].classList.contains('js-child-has-focus')).toBe(false);
```

You could also check it hasn't mashed the classes together (like it did before).

File: tests/javascripts/updateContent.test.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/javascripts/updateContent.test.js]: I don't think it's changed really: my concern was more with the assertion itself e.g. a typo of `dta-resource` would always be false. But it does look right just from eyeballing it, so let's leave as-is.

**Review by benthorner — APPROVED** [review]
✅ 

Thanks again for all the chats around this and being open to all kinds of comments from me 🏅. I'll leave you with two optional comments (the unresolved ones) 👋.

File: tests/javascripts/updateContent.test.js (hunk 11 above)
Comment (tombye) [reply tests/javascripts/updateContent.test.js:492]: I like this point, and would have just commited your suggestion but I wanted to check no classes were leaking either way so added another check to this here: https://github.com/alphagov/notifications-admin/pull/4155/commits/8521d1e45f43cdbe7a07e0e988ddb6b1f7ac868a

Comment (benthorner) [reply tests/javascripts/updateContent.test.js:492]: Nice, looks good to me 👍.

**Comment by tombye** [comment]
Thanks for this @benthorner, appreciate the review and associated chats. The later, I think, were really useful for keeping some of the discussions in scope so much appreciated.
